<!-- Turtlebot3 -->

<launch>
  <!--- Robot -->
  <include file="$(find turtlebot3_bringup)/launch/includes/description.launch.xml"/>
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
    <param name="publish_frequency" type="double" value="50.0" />
  </node>

  <!-- Map server -->
  <arg name="map_file" default="$(find slam_navi)/maps/map.yaml"/>
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)">
  </node>

  <!--- Depth image to laser scan -->
  <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan" >
    <param name="scan_height" value="1"/>
    <param name="range_min" value="0.1"/>
    <param name="range_max" value="10"/>
    <param name="output_frame_id" value="base_link"/>
    <remap from="image" to="camera/depth/image_raw"/>
    <remap from="scan" to="scan/depth"/>
    <param name="scan_time" value="10.0"/>
  </node>

  <!-- AMCL -->
  <node pkg="amcl" type="amcl" name="amcl">
    <param name="min_particles"             value="500"/>
    <param name="max_particles"             value="3000"/>
    <param name="kld_err"                   value="0.02"/>
    <param name="update_min_d"              value="0.20"/>
    <param name="update_min_a"              value="0.20"/>
    <param name="resample_interval"         value="1"/>
    <param name="transform_tolerance"       value="0.5"/>
    <param name="recovery_alpha_slow"       value="0.00"/>
    <param name="recovery_alpha_fast"       value="0.00"/>
    <param name="initial_pose_x"            value="0.0"/>
    <param name="initial_pose_y"            value="0.0"/>
    <param name="initial_pose_a"            value="0.0"/>
    <param name="gui_publish_rate"          value="50.0"/>
    <param name="laser_max_range"           value="3.5"/>
    <param name="laser_max_beams"           value="180"/>
    <param name="laser_z_hit"               value="0.5"/>
    <param name="laser_z_short"             value="0.05"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.5"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="laser_model_type"          value="likelihood_field"/>
    <param name="odom_model_type"           value="diff"/>
    <param name="odom_alpha1"               value="0.1"/>
    <param name="odom_alpha2"               value="0.1"/>
    <param name="odom_alpha3"               value="0.1"/>
    <param name="odom_alpha4"               value="0.1"/>
    <param name="odom_frame_id"             value="odom"/>
    <param name="base_frame_id"             value="base_footprint"/>
  </node>
 
  <!-- Move_Base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    <rosparam file="$(find slam_navi)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find slam_navi)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find slam_navi)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find slam_navi)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find slam_navi)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find slam_navi)/param/dwa_local_planner_params.yaml" command="load" />
    <remap from="cmd_vel" to="/mobile_base/commands/velocity"/>
  </node>

  <!-- run RViz node (visualization) -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find slam_navi)/rviz/navi.rviz">
  </node>
  
  <!-- Darknet -->
  <include file="$(find darknet_ros)/launch/darknet_ros.launch"/>
  
  <!-- reinforcement -->
  <node pkg="turtlebot3_navigation" type="reinforcement.py" name="reinforcement" output="screen">
  <param name="map_topic" value="/map"/>
  <param name="rate" value="100"/>
  </node>

</launch>
